mypy/subtypes.py

```diff
@@ -1008,10 +1007,15 @@ def unify_generic_callable(type: CallableType, target: CallableType,
     if None in inferred_vars:
         return None
     non_none_inferred_vars = cast(List[Type], inferred_vars)
-    msg = messages.temp_message_builder()
-    applied = mypy.applytype.apply_generic_arguments(type, non_none_inferred_vars, msg,
+    had_errors = False
+
+    def report(*args: Any) -> None:
+        nonlocal had_errors
+        had_errors = True
+
+    applied = mypy.applytype.apply_generic_arguments(type, non_none_inferred_vars, report,
                                                      context=target)
-    if msg.is_errors():
```

Comment:
This pattern is unfortunately used quite often in other contexts (in particular in overloads, see https://github.com/python/mypy/issues/7367). It would be great to get rid of those (but in case of methods I would probably prefer `self.flag` rather that `nonlocal`).